crates/parser/src/format/character_separated/mod.rs

```diff
@@ -67,74 +52,125 @@ fn detect_quote_char(delimiter: u8, peeked: &[u8]) -> Option<u8> {
         None
     } else if n_double >= n_single {
         // If there's a tie, then I guess double quotes win?
-        Some(b'"')
+        Some(Quote::DoubleQuote)
     } else {
-        Some(b'\'')
+        Some(Quote::SingleQuote)
+    }
+}
+
+/// Applies a not-quite-best-effort heuristic to determine a delimiter character that can hopefully
+/// work to parse the CSV file that has a prefix of `peeked` bytes. This implementation is super
+/// basic. It just counts the occurrences of common delimiters and returns the one with the most.
+/// A more sophisticated method would be to split on the line ending
+fn detect_delimiter(peeked: &[u8]) -> Delimiter {
+    let mut delims = Vec::<(Delimiter, usize)>::new();
+    for candidate in Delimiter::iter() {
+        let n = bytecount::count(peeked, candidate.byte_value());
+        delims.push((candidate, n));
     }
+
+    // Sort by the number of observed matches, and then by the byte value of the delimiter. This
+    // means that lower numbered delimiters, such as control characters, will take precedence over
+    // those with higher byte values such as printable characters. This only serves to break a tie
+    // in the number of observed values, though, so that the selected delimiter in such cases is
+    // deterministic.
+    delims.sort_by_key(|elem| (peeked.len() - elem.1 as usize, elem.0.byte_value()));
+
+    let best = delims.first().unwrap();
+    // TODO: consider returning an error if the best delimiter still has 0 occurrences
```

Comment:
`delims` will contain an entry for each possible delimiter, even if no characters for it were actually observed. Because of the sorting behavior, this means that if we don't observe any of the possible delimiters, then we'll actually end up picking `0x00` as the delimiter. While the `0x00` is somewhat arbitrary, it seems sufficient to properly parse CSVs that have only a single column, which wouldn't normally contain any delimiters. This case of single-column CSVs is what motivated the behavior here. I pushed a commit to add a comment and a test case to make this explicit, and also to remove that bogus TODO.